pilot/pkg/model/sidecar.go

```diff
@@ -0,0 +1,319 @@
+	}
+	defaultEgressListener.importedServices = defaultEgressListener.selectServices(ps.Services(&dummyNode))
+
+	out := &SidecarScope{
+		EgressListeners:             []*IstioListenerWrapper{defaultEgressListener},
+		allImportedServices:         defaultEgressListener.importedServices,
+		allImportedDestinationRules: make(map[Hostname]*Config),
+	}
+
+	// Now that we have all the services that sidecars using this scope (in
+	// this config namespace) will see, identify all the destinationRules and virtualServices
+	// that these services need
+	for _, s := range out.allImportedServices {
+		out.allImportedDestinationRules[s.Hostname] = ps.DestinationRule(&dummyNode, s.Hostname)
+	}
+
+	meshGateway := map[string]bool{IstioMeshGateway: true}
+	out.allImportedVirtualServices = out.selectVirtualServices(ps.VirtualServices(&dummyNode, meshGateway))
+	return out
+}
+
+// ConvertToSidecarScope converts from Sidecar config to SidecarScope object
+func ConvertToSidecarScope(ps *PushContext, sidecarConfig *Config) *SidecarScope {
+	out := &SidecarScope{
+		Config:                      sidecarConfig,
+		allImportedServices:         make([]*Service, 0),
+		allImportedDestinationRules: make(map[Hostname]*Config),
+	}
+
+	r := sidecarConfig.Spec.(*networking.Sidecar)
+	out.EgressListeners = make([]*IstioListenerWrapper, 0)
+	out.IngressListeners = make([]*IstioListenerWrapper, 0)
+
+	for _, e := range r.Egress {
+		out.EgressListeners = append(out.EgressListeners, convertIstioListenerToWrapper(ps, sidecarConfig, e))
+	}
+
+	for _, e := range r.Ingress {
+		// TODO: These need to go into CDS as well
+		out.IngressListeners = append(out.IngressListeners, convertIstioListenerToWrapper(ps, sidecarConfig, e))
+	}
+
+	// Now collect all the imported services across all egress listeners. This is needed to generate CDS output
+	servicesAdded := make(map[Hostname]struct{})
+	dummyNode := Proxy{
+		ConfigNamespace: sidecarConfig.Namespace,
+	}
+
+	for _, listener := range out.EgressListeners {
+		for _, s := range listener.importedServices {
+			// TODO: port merging when each listener generates a partial service
+			if _, found := servicesAdded[s.Hostname]; !found {
+				servicesAdded[s.Hostname] = struct{}{}
+				out.allImportedServices = append(out.allImportedServices, s)
+			}
+		}
+	}
+
+	// Now that we have all the services that sidecars using this scope (in
+	// this config namespace) will see, identify all the destinationRules and virtualServices
+	// that these services need
+	for _, s := range out.allImportedServices {
+		out.allImportedDestinationRules[s.Hostname] = ps.DestinationRule(&dummyNode, s.Hostname)
+	}
+	meshGateway := map[string]bool{IstioMeshGateway: true}
+	out.allImportedVirtualServices = out.selectVirtualServices(ps.VirtualServices(&dummyNode, meshGateway))
+	return out
+}
+
+// Services returns the list of services imported across all egress listeners by this
+// Sidecar config
+func (sc *SidecarScope) Services() []*Service {
```

Comment:
Pass a proxy as parameter - I assume the intent is to move the code that gets services here - but it'll need to be 
per proxy with on-demand. See the comments above.

Reply:
This is per sidecar Scope object. Any proxy that selects this sidecar scope object will get the services attached to this Sidecarscope object.

Reply:
Then don't call this method when generating the config - and call a method that takes proxy, and combine the struct that you are removing with the sidecar services. 

Or just change the UpdateNoteIsolation to add the SidecarScope.Services to the per-proxy struct.

Reply:
I'll make this change in a separate PR, resolving this.